cli/report/report.py

```diff
@@ -176,6 +181,58 @@ def file_jira_issues(
 
         return bugs_filed
 
+    def filter_priority_rule_failure_pairs(
+        self,
+        rule_failure_pairs: list[dict[Any, Any]],
+    ) -> list[dict[Any, Any]]:
+        """
+        Filters a list of rule/failure pairs based on user-defined groups/priorities. If there are multiple rule/failure
+        pairs with a group defined in the rule, it will determine which rule is the highest priority. After filtering
+        out the lower priority rule/failure pairs, it will return the new list of rule/failure pairs.
+
+        Args:
+            rule_failure_pairs (list[dict[Any, Any]]): A list of rule/failure pairs.
+        Returns:
+            list[dict[Any, Any]]: A list of rule/failure pairs that have been filtered for user-defined priorities.
+        """
+
+        groups: dict[Any, Any] = {}  # {"some-group-name": [pair1, pair2, pair3]}
+        filtered_rule_failure_pairs = []
+
+        # Populate groups
+        for pair in rule_failure_pairs:
+            if pair["rule"].group_name and pair["rule"].group_priority:
+                # Add the group if it isn't in groups dictionary
+                if pair["rule"].group_name not in groups:
+                    groups.update({pair["rule"].group_name: []})
+                # Add the pair to list of groups
+                groups[pair["rule"].group_name].append(pair)
+            else:
+                filtered_rule_failure_pairs.append(pair)
+
+        # Find the highest priority pair for each "group" item. This should make the list of pairs for each group
+        # item is only 1 item long.
+        for group in groups:
```

Comment:
What about checking that `priority` is unique?
User can send
```
"group": {"name": "some-group", "priority": 1}
"group": {"name": "some-group", "priority": 2}
"group": {"name": "some-group", "priority": 1}
```
Do we want to exit in such case?

Reply:
I can add some validation to the methods I use in rule.py to get those values from the user's config.

Reply:
@myakove Your suggestion in [this comment](https://github.com/CSPI-QE/firewatch/pull/42#discussion_r1309931610) actually causes this to report on both of the steps with `"priority": 1`. I think I like that behavior.